modules/preloader/injected/EthereumProvider.js

```diff
@@ -1,191 +1,201 @@
+const origin = this.origin;
+
 (function() {
-  var EventEmitter = window.EventEmitter;
+  class EthereumProvider extends window.EventEmitter {
+    constructor() {
+      // Call super for `this` to be defined
+      super();
+
+      // Init storage
+      this._isConnected = false;
+      this._nextJsonrpcId = 0;
+      this._promises = {};
+      this._activeSubscriptions = [];
+
+      // Fire the connect
+      this._connect();
+
+      // Listen for jsonrpc responses
+      window.addEventListener('message', this._handleJsonrpcMessage.bind(this));
+    }
+
+    /* Methods */
 
-  var postMessage = function(payload) {
-    if (typeof payload === 'object') {
-      payload = JSON.stringify(payload);
+    enable() {
+      return new Promise((resolve, reject) => {
+        window.mist
+          .requestAccounts()
+          .then(resolve)
+          .catch(reject);
+      });
     }
 
-    window.postMessage(
-      payload,
-      !location.origin || location.origin === 'null' ? '*' : location.origin
-    );
-  };
+    send(method, params = []) {
+      if (!method || typeof method !== 'string') {
+        return new Error('Method is not a valid string.');
+      }
+
+      if (!(params instanceof Array)) {
+        return new Error('Params is not a valid array.');
+      }
+
+      const id = this._nextJsonrpcId++;
+      const jsonrpc = '2.0';
+      const payload = { jsonrpc, id, method, params };
+
+      const promise = new Promise((resolve, reject) => {
+        this._promises[payload.id] = { resolve, reject };
+      });
+
+      // Send jsonrpc request to node
+      window.postMessage(
+        { type: 'mistAPI_ethereum_provider_write', message: payload },
+        origin
+      );
+
+      return promise;
+    }
 
-  // on events are: "connect", "data", "error", "end", "timeout"
-  // "data" will get notifications
+    subscribe(subscriptionType, params) {
+      return this.send('eth_subscribe', [subscriptionType, ...params]).then(
```

Comment:
As I see here is the ```eth_subscribe``` method hardcoded in the subscribe method of the provider.. is this correct that whisper subscriptions wont be possible with the ethereumProvider? (whisper use ```shh_subscribe``` as json-rpc method)